plugins/processors/execd/execd.go

```diff
@@ -0,0 +1,147 @@
+package execd
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"log"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/config"
+	"github.com/influxdata/telegraf/internal/process"
+	"github.com/influxdata/telegraf/plugins/parsers"
+	"github.com/influxdata/telegraf/plugins/processors"
+	"github.com/influxdata/telegraf/plugins/serializers"
+)
+
+const sampleConfig = `
+  ## Program to run as daemon
+  command = ["telegraf-smartctl", "-d", "/dev/sda"]
+
+  ## Delay before the process is restarted after an unexpected termination
+  restart_delay = "10s"
+`
+
+type Execd struct {
+	Command      []string
+	RestartDelay config.Duration
+
+	parserConfig     *parsers.Config
+	parser           parsers.Parser
+	serializerConfig *serializers.Config
+	serializer       serializers.Serializer
+	acc              telegraf.Accumulator
+	process          *process.Process
+}
+
+func New() *Execd {
+	return &Execd{
+		RestartDelay: config.Duration(10 * time.Second),
+		parserConfig: &parsers.Config{
+			DataFormat: "influx",
+		},
+		serializerConfig: &serializers.Config{
+			DataFormat: "influx",
+		},
+	}
+}
+
+func (e *Execd) SampleConfig() string {
+	return sampleConfig
+}
+
+func (e *Execd) Description() string {
+	return "Run executable as long-running processor plugin"
+}
+
+func (e *Execd) Start(acc telegraf.Accumulator) error {
+	var err error
+	e.parser, err = parsers.NewParser(e.parserConfig)
+	if err != nil {
+		return fmt.Errorf("error creating parser: %w", err)
+	}
+	e.serializer, err = serializers.NewSerializer(e.serializerConfig)
+	if err != nil {
+		return fmt.Errorf("error creating serializer: %w", err)
+	}
+	e.acc = acc
+
+	if len(e.Command) == 0 {
+		return fmt.Errorf("FATAL no command specified")
+	}
+
+	e.process, err = process.New(e.Command)
+	if err != nil {
+		return fmt.Errorf("Error creating new process: %w", err)
+	}
+
+	e.process.RestartDelay = time.Duration(e.RestartDelay)
+	e.process.ReadStdoutFn = e.cmdReadOut
+	e.process.ReadStderrFn = e.cmdReadErr
+
+	if err = e.process.Start(); err != nil {
+		return fmt.Errorf("failed to start process %s: %w", e.Command, err)
+	}
+
+	return nil
+}
+
+func (e *Execd) Add(m telegraf.Metric, acc telegraf.Accumulator) {
+	b, err := e.serializer.Serialize(m)
+	if err != nil {
+		log.Println(fmt.Errorf("Metric serializing error: %s", err))
+		return
+	}
+	// if osStdin, ok := e.process.Stdin.(*os.File); ok {
+	// 	osStdin.SetWriteDeadline(time.Now().Add(1 * time.Second))
+	// }
+
+	_, err = e.process.Stdin.Write(b)
```

Comment:
In this processor we cannot maintain tracking metrics, because input/output is done asynchronously and we don't have any other way to tie the input to output.

Let's call metric.Drop() on every metric and document that tracking is removed in the README.

We can try to handle the tracking metrics in the future with the more tightly integrated process plugin using a different ipc method.

Reply:
👍 a case for metric metadata :D